Lib/test/test_listcomps.py

```diff
@@ -379,6 +379,23 @@ def f():
         with self.assertRaises(UnboundLocalError):
             f()
 
+    def test_unbound_local_in_class_scope(self):
+        class X:
+            y = 1
+            with self.assertRaises(NameError):
+                [x + y for x in range(2)]
+
+    def test_comprehension_in_class_scope(self):
+        code = """
+            y = 1
+            class X:
+                y = 2
+                vals = [(x, y) for x in range(2)]
+            vals = X.vals
+        """
+        self._check_in_scopes(code, {"vals": [(0, 1), (1, 1)]},
+                              scopes=["module", "fucntion"])
+
```

Comment:
Maybe take a look at the tests I have on https://github.com/python/cpython/pull/104519 and let's try to get the union of all the cases covered.